proof.go

```diff
@@ -465,6 +465,203 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
+// More information on the algorithm used can be found in the ToLeafRanges() method docs.
+func (proof Proof) VerifySubtreeRootInclusion(nth *NmtHasher, subtreeRoots [][]byte, subtreeWidth int, root []byte) (bool, error) {
+	// check that the proof range is valid
+	if proof.Start() < 0 || proof.Start() >= proof.End() {
+		return false, fmt.Errorf("proof range [proof.start=%d, proof.end=%d) is not valid: %w", proof.Start(), proof.End(), ErrInvalidRange)
+	}
+
+	// check that the root is valid w.r.t the NMT hasher
+	if err := nth.ValidateNodeFormat(root); err != nil {
+		return false, fmt.Errorf("root does not match the NMT hasher's hash format: %w", err)
+	}
+	// check that all the proof.Notes() are valid w.r.t the NMT hasher
+	for _, node := range proof.Nodes() {
+		if err := nth.ValidateNodeFormat(node); err != nil {
+			return false, fmt.Errorf("proof nodes do not match the NMT hasher's hash format: %w", err)
+		}
+	}
+	// check that all the subtree roots are valid w.r.t the NMT hasher
+	for _, subtreeRoot := range subtreeRoots {
+		if err := nth.ValidateNodeFormat(subtreeRoot); err != nil {
+			return false, fmt.Errorf("inner nodes does not match the NMT hasher's hash format: %w", err)
+		}
+	}
+
+	// get the subtree roots leaf ranges
+	ranges, err := ToLeafRanges(proof.Start(), proof.End(), subtreeWidth)
+	if err != nil {
+		return false, err
+	}
+
+	// check whether the number of ranges matches the number of subtree roots.
+	// if not, make an early return.
+	if len(subtreeRoots) != len(ranges) {
+		return false, fmt.Errorf("number of subtree roots %d is different than the number of leaf ranges %d", len(subtreeRoots), len(ranges))
+	}
+
+	var computeRoot func(start, end int) ([]byte, error)
+	// computeRoot can return error iff the HashNode function fails while calculating the root
+	computeRoot = func(start, end int) ([]byte, error) {
+		// reached a leaf
+		if end-start == 1 {
+			// if the leaf index falls within the proof range, pop and return a
+			// leaf
+			if proof.Start() <= start && start < proof.End() {
+				// this check should always be the case.
+				// however, it is added to avoid nil pointer exceptions
+				if len(ranges) != 0 {
+					// advance the list of ranges
+					ranges = ranges[1:]
+				}
+				// advance leafHashes
+				return popIfNonEmpty(&subtreeRoots), nil
+			}
+
+			// if the leaf index is outside the proof range, pop and return a
+			// proof node (which in this case is a leaf) if present, else return
+			// nil because leaf doesn't exist
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		// if the current range does not overlap with the proof range, pop and
+		// return a proof node if present, else return nil because subtree
+		// doesn't exist
+		if end <= proof.Start() || start >= proof.End() {
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
+			ranges = ranges[1:]
+			return popIfNonEmpty(&subtreeRoots), nil
+		}
+
```

Comment:
Note: 
This could be refactored to be this:
```
		// if the current range has a subtree root, then we use it
		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
			ranges = ranges[1:]
			return popIfNonEmpty(&subtreeRoots), nil
		}

		// reached a leaf
		if end-start == 1 {
			// if the leaf index is outside the proof range, pop and return a
			// proof node (which in this case is a leaf) if present, else return
			// nil because leaf doesn't exist
			return popIfNonEmpty(&proof.nodes), nil
		}

		// if the current range does not overlap with the proof range, pop and
		// return a proof node if present, else return nil because subtree
		// doesn't exist
		if end <= proof.Start() || start >= proof.End() {
			return popIfNonEmpty(&proof.nodes), nil
		}
```

and I tried testing it and it works fine. However, I would like a second confirmation that nothing will break with this refactor

Reply:
Looks correct to me, too. Would still be good to closely look into the case `if end-start == 1` and why both related code blocks are equivalent in all (edge) cases.

Reply:
https://github.com/celestiaorg/nmt/issues/263

Reply:
I'd even say you can revise the refactored version further (as suggested in this [comment](https://github.com/celestiaorg/nmt/pull/260#discussion_r1653378546) as well), and it should work:
```
		// if the current range has a subtree root, then we use it
		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
			ranges = ranges[1:]
			return popIfNonEmpty(&subtreeRoots), nil
		}

		// if the current range does not overlap with the proof range, pop and
		// return a proof node if present, else return nil because subtree
		// doesn't exist
		if end <= proof.Start() || start >= proof.End() {
			return popIfNonEmpty(&proof.nodes), nil
		}
```
Rationale: When traversing the tree from top (root) to bottom (leaves), we split the range covered by the root to half, and for each half i.e., subtree there can be 3 cases:
- The subtree range exactly matches one of the supplied subtree roots -> hence return that subtree root
- The subtree range overlaps with the proof range but does not exactly match the range of any of the subtree roots yet -> keep splitting that subtree 
- The subtree range is entirely outside of the proof range -> there needs to be a node for that in the proof, hence pop and return it

Reply:
Nevertheless, even if you want to keep your refactored version with the additional `if end-start == 1 {` block in it, I won't oppose it, though; one thing you could do is to refactor it slightly differently:
```
		// if the current range has a subtree root, then we use it
		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
			ranges = ranges[1:]
			return popIfNonEmpty(&subtreeRoots), nil
		}

		// if the current range does not overlap with the proof range, pop and
		// return a proof node if present, else return nil because subtree
		// doesn't exist
		if end <= proof.Start() || start >= proof.End() {
			return popIfNonEmpty(&proof.nodes), nil
		}

		// reached a leaf
		if end-start == 1 {
			// if the leaf index is outside the proof range, pop and return a
			// proof node (which in this case is a leaf) if present, else return
			// nil because leaf doesn't exist
			return popIfNonEmpty(&proof.nodes), nil
		}

```
 In this new version, I doubt we can find any input that can make that `if end-start == 1 ` block to be executed at all. All the inputs will enter either the first `if` or the second one, but never the third (the second if covers the third one already).

Reply:
One more point, in the refactored version you [shared](https://github.com/celestiaorg/nmt/pull/260#discussion_r1624770516), i.e, this if specifically 
```
               // if the current range has a subtree root, then we use it
		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
			ranges = ranges[1:]
			return popIfNonEmpty(&subtreeRoots), nil
		}
```
I think it needs to be revised. First  we need to check whether the proof range proof.Start(),Proof.End() contains the current range i.e., [start,end), and if it was the case then check if `if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end `, and if this was the case, we return a subtreeRoot, otherwise, this is an error and an error needs to be returned.

Reply:
@staheri14 Why is it an error? if I understand correctly, if we have for example the proof for the range `[0, 8)` in a `[0,15]` tree.
We could be at level `[4,8)`, for example, and this range is contained in the proof range, however, there is no subtree root for that given the subtree root width is 0. 

Separately, I removed the whole block:

```golang
if end-start == 1 {
			// if the leaf index falls within the proof range, pop and return a
			// leaf
			if proof.Start() <= start && start < proof.End() {
				// this check should always be the case.
				// however, it is added to avoid nil pointer exceptions
				if len(ranges) != 0 {
					// advance the list of ranges
					ranges = ranges[1:]
				}
				// advance leafHashes
				return popIfNonEmpty(&subtreeRoots), nil
			}
			// if the leaf index is outside the proof range, pop and return a
			// proof node (which in this case is a leaf) if present, else return
			// nil because leaf doesn't exist
			return popIfNonEmpty(&proof.nodes), nil
		}
```

and ran the all combinations test and it works fine. So probably we can remove it no problem. And we end up with:
```golang
	computeRoot = func(start, end int) ([]byte, error) {
		if len(ranges) != 0 && ranges[0].Start == start && ranges[0].End == end {
			ranges = ranges[1:]
			return popIfNonEmpty(&subtreeRoots), nil
		}

		// if the current range does not overlap with the proof range, pop and
		// return a proof node if present, else return nil because subtree
		// doesn't exist
		if end <= proof.Start() || start >= proof.End() {
			return popIfNonEmpty(&proof.nodes), nil
		}

		// Recursively get left and right subtree
		k := getSplitPoint(end - start)
		left, err := computeRoot(start, start+k)
		if err != nil {
			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start, start+k, err)
		}
		right, err := computeRoot(start+k, end)
		if err != nil {
			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start+k, end, err)
		}

		// only right leaf/subtree can be non-existent
		if right == nil {
			return left, nil
		}
		hash, err := nth.HashNode(left, right)
		if err != nil {
			return nil, fmt.Errorf("failed to hash node: %w", err)
		}
		return hash, nil
	}
```

Also, we can remove the power of 2 constraint on the tree size.

if you agree, I'll go ahead and commit these changes.

Reply:
I also added another [comment](https://github.com/celestiaorg/nmt/pull/260/files#r1676117495)  which also applies to your new suggested code block.